File: 1-EASY/Python/3-fibonacci-sequence-generator/proposed-solution/main.py
# Challenge: Create a generator that returns the Fibonacci sequence up to n terms.
# You can explore iterative and recursive approaches to solve the problem.

def fibonacci_iterative(n):
    a, b = 0, 1
    for _ in range(n):
        yield a
        a, b = b, a + b

def fibonacci_recursive(n, a=0, b=1):
    if n == 0:
        return
    yield a
    yield from fibonacci_recursive(n-1, b, a + b)

# Example usage
if __name__ == "__main__":
    n = 10
    print("Iterative Fibonacci sequence:")
    for num in fibonacci_iterative(n):
        print(num, end=" ")
    print("\nRecursive Fibonacci sequence:")
    for num in fibonacci_recursive(n):
        print(num, end=" ")
    print()

File: 1-EASY/JavaScript/3-fibonacci-sequence-generator/proposed-solution/main.js
// Challenge: Create a generator that returns the Fibonacci sequence up to n terms.
// You can explore iterative and recursive approaches to solve the problem.

// Iterative generator function
function* fibonacciIterative(n) {
    let a = 0, b = 1;
    for (let i = 0; i < n; i++) {
        yield a;
        [a, b] = [b, a + b];
    }
}

// Recursive generator function
function* fibonacciRecursive(n, a = 0, b = 1) {
    if (n === 0) return;
    yield a;
    yield* fibonacciRecursive(n - 1, b, a + b);
}

// Example usage
const n = 10;
console.log("Iterative Fibonacci sequence:");
for (const num of fibonacciIterative(n)) {
    process.stdout.write(num + " ");
}
console.log("\nRecursive Fibonacci sequence:");
for (const num of fibonacciRecursive(n)) {
    process.stdout.write(num + " ");
}
console.log();
